.github/workflows/build.yml

```diff
@@ -4,128 +4,96 @@ name: 'Build'
 on: ['push', 'pull_request']
 
 jobs:
-  test:
-    name: 'Test'
+  # Build the RocksDB C library and cache the result.
+  librocksdb:
+    name: 'Build librocksdb'
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
         os: [ubuntu-latest]
-          #py_ver: ['3.7', '3.8', '3.9']
-          #rocksdb_ver: ['v6.14.6', 'v6.11.4', 'v5.17.2']
-        py_ver: ['3.9']
-        rocksdb_ver: ['v6.14.6']
+        rocksdb_ver: ['v6.14.6', 'v6.11.4']
 
     steps:
-      - uses: actions/checkout@v2
-        name: 'Checkout source repository'
-
-      - uses: actions/setup-python@v2
-        name: 'Set up Python ${{ matrix.py_ver }}'
+      - uses: actions/cache@v2
+        id: cache-librocksdb
         with:
-          python-version: ${{ matrix.py_ver }}
+          key: ${{ matrix.os }}-librocksdb-${{ matrix.rocksdb_ver }}
+          path: /opt/rocksdb
+          restore-keys: |
+            librocksdb-${{ matrix.os }}-
+            librocksdb-
 
       - name: 'Install dependencies'
-        run: |
-          sudo apt install -y libsnappy-dev libbz2-dev liblz4-dev libz-dev libgflags-dev libzstd-dev
+        if: steps.cache-librocksdb.outputs.cache-hit != 'true'
+        run: >
+          sudo apt install -y libsnappy-dev libbz2-dev liblz4-dev libz-dev
+          libgflags-dev libzstd-dev
 
-      - name: 'Install RocksDB ${{ matrix.rocksdb_ver }}'
+      - name: 'Clone & build RocksDB ${{ matrix.rocksdb_ver }}'
+        if: steps.cache-librocksdb.outputs.cache-hit != 'true'
         run: |
           pushd /opt &&
           git clone https://github.com/facebook/rocksdb &&
           cd rocksdb &&
           git reset --hard ${{ matrix.rocksdb_ver }} &&
           CXXFLAGS='-flto -Os -s' PORTABLE=1 make shared_lib -j 4 &&
```

Comment:
I haven't been able to test yet (I tried #3 for that) but, if there's a partial key match (say, for a different RocksDB version in the same operating system)… I _fear_ the git clone might fail, because the directory eists and is not empty?

Plus, what is the benefit of a partial-key match here? We could reuse the clone, but it'd also have the artifacts of a previous build, and no (visible) clean is being done. So after the reset, make would have to get right to redo all targets, correctly. (Of course that its job, but some dependency might have been missed.)

I _think_ I'd remove partial-key matches completely and, perhaps, use the standard checkout@v2 action to checkout the repository (not sure if that takes a destination_path: /opt). Or, at least, perhaps use `--depth 0 --branch ${{ rocksdb_ver }}`?

Reply:
You are right, I completely missed this. I checked, and the checkout action does not let you specify a path outside of the workdir, so I kept the manual clone, but changed it to avoid reusing a cache from a different version.